k8s/src/main/scala/io/buoyant/k8s/istio/IdentifierPreconditions.scala

```diff
@@ -23,32 +36,57 @@ trait IdentifierPreconditions {
     }
   }
 
-  def matchesAllConditions(getHeader: (String) => Option[String], matchCondition: MatchCondition): Boolean = {
+  def matchesAllConditions(req: IstioRequestMeta, matchCondition: MatchCondition): Boolean = {
     val matchesHeaders = matchCondition.`httpHeaders`.forall {
+      case ("uri", m) => headerMatches(req.uri, m)
+      case ("scheme", m) => headerMatches(req.scheme, m)
+      case ("method", m) => headerMatches(req.method, m)
+      case ("authority", m) => headerMatches(req.authority, m)
       case (headerName, stringMatch) =>
-        // return false if the headerName does not appear in the request's headerMap
-        getHeader(headerName) match {
-          case Some(headerValue) => headerMatches(headerValue, stringMatch)
+        req.getHeader(headerName) match {
+          case Some(a) => headerMatches(a, stringMatch)
           case None => false
         }
     }
     //TODO: add other match conditions
     matchesHeaders
   }
 
-  def filterRules(rules: Map[String, RouteRule], dest: String, getHeader: (String) => Option[String]): Seq[(String, RouteRule)] = rules.filter {
+  /**
+   * Rewrites uri and authority based on a route-rule. It doesn't modify
+   * the request directly because it's used by both http and h2.
+   * @param rule
+   * @param uri
+   * @param authority
+   * @return Tuple of uri and authority to used to rewrite headers on the request
+   */
+  def httpRewrite(rule: RouteRule, uri: String, authority: Option[String]): (String, Option[String]) = {
+    rule.`rewrite` match {
+      case Some(HTTPRewrite(url, updatedAuth)) =>
+        val updatedUri = url.map { replacement =>
+          rule.`match`.flatMap(_.`httpHeaders`.get("uri").flatMap(_.`matchType`)) match {
+            case Some(OneofMatchType.Prefix(pfx)) if uri.startsWith(pfx) => uri.replace(pfx, replacement)
+            case _ => replacement
```

Comment:
Not sure if the istio spec spells out what this behavior should be if the match type is prefix but the prefix doesn't match.  I would expect the uri to be left unmodified, rather than replaced.